bentoml/yatai/deployment/azure_functions/operator.py

```diff
@@ -0,0 +1,705 @@
+# Copyright 2019 Atalaya Tech, Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+# http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import json
+import os
+import re
+import shutil
+import subprocess
+import logging
+import sys
+
+import docker
+
+from bentoml.utils.tempdir import TempDirectory
+from bentoml.saved_bundle import loader
+from bentoml.yatai.deployment.azure_functions.templates import AZURE_API_FUNCTION_JSON
+from bentoml.yatai.deployment.operator import DeploymentOperatorBase
+from bentoml.yatai.deployment.utils import ensure_docker_available_or_raise
+from bentoml.exceptions import (
+    BentoMLException,
+    MissingDependencyException,
+    AzureServiceError,
+    YataiDeploymentException,
+)
+from bentoml.yatai.proto.deployment_pb2 import (
+    ApplyDeploymentResponse,
+    DeploymentState,
+    DescribeDeploymentResponse,
+    DeleteDeploymentResponse,
+)
+from bentoml.yatai.proto.repository_pb2 import GetBentoRequest
+from bentoml.yatai.status import Status
+from bentoml.configuration import LAST_PYPI_RELEASE_VERSION
+
+logger = logging.getLogger(__name__)
+
+
+MAX_RESOURCE_GROUP_NAME_LENGTH = 90
+MAX_STORAGE_ACCOUNT_NAME_LENGTH = 24
+MAX_FUNCTION_NAME_LENGTH = 60
+MAX_CONTAINER_REGISTRY_NAME_LENGTH = 50
+
+# https://docs.microsoft.com/en-us/azure/azure-functions/functions-premium-plan
+AZURE_FUNCTIONS_PREMIUM_PLAN_SKUS = ['EP1', 'EP2', 'EP3']
+# https://docs.microsoft.com/en-us/java/api/com.microsoft.azure.functions.annotation.httptrigger.authlevel?view=azure-java-stable
+AZURE_FUNCTIONS_AUTH_LEVELS = ['anonymous', 'function', 'admin']
+DEFAULT_MIN_INSTANCE_COUNT = 1
+DEFAULT_MAX_BURST = 20
+DEFAULT_PREMIUM_PLAN_SKU = AZURE_FUNCTIONS_PREMIUM_PLAN_SKUS[0]
+DEFAULT_FUNCTION_AUTH_LEVEL = AZURE_FUNCTIONS_AUTH_LEVELS[0]
+
+
+def _assert_az_cli_logged_in():
+    account_info = _call_az_cli(
+        command=['az', 'account', 'show'], message='show Azure account'
+    )
+    if not account_info['user'] or not account_info['homeTenantId']:
+        raise YataiDeploymentException(
+            'A signed in Azure CLI is required for Azure Function deployment'
+        )
+
+
+def _assert_azure_cli_available():
+    try:
+        _call_az_cli(command=['az', 'account', 'show'], message='show Azure account')
```

Comment:
might be better to use something like `az --help` for this?  things like `account show` probably goes to audit logs and might not be necessary for us to do this operation here anyway